extends PanelContainer

class_name Player

signal health_changed(old_value, new_value)
signal health_depleted

var health: int = 100:
	set(value):
		health = value
		update_unit_info()

## –ò–º—è —é–Ω–∏—Ç–∞. –ù—É–∂–Ω–æ –ø–µ—Ä–µ–¥–µ–ª–∞—Ç—å –Ω–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç
@export var unit_name:String = '–Æ–Ω–∏—Ç'

## –ù–∞—á–∞–ª—å–Ω—ã–µ —Å—Ç–∞—Ç—ã —é–Ω–∏—Ç–∞
@export var STATS: StartingStats

## –£–º–µ–Ω–∏—è —é–Ω–∏—Ç–∞
@export var SKILLS: Array[Skill]

## –ö–æ–º–ø–æ–Ω–µ–Ω—Ç—ã —é–Ω–∏—Ç–∞
@export var speed_component: SpeedComponent
@export var active_component: ActiveComponent
@onready var level_component: LevelComponent = $LevelComponent
@onready var mastery_component: MasteryComponent = $MasteryComponent
@onready var skills_component: SkillsComponent = $SkillsComponent

## –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è —Å—Ç–∞—Ç–æ–≤ —é–Ω–∏—Ç–∞ –∏ –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏–∏ –æ –Ω—ë–º
func _ready():
	speed_component.max_speed = STATS.speed
	speed_component.speed = STATS.speed
	%Sprite2D.texture = STATS.texture
	update_unit_info()
	
## –û–±–Ω–æ–≤–ª–µ–Ω–∏–µ –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏–∏ –æ –≥–µ—Ä–æ–µ
func update_unit_info():
	%InfoPanel/Name.text = "%s. –£—Ä. %s" % [unit_name, level_component.level]
	%HelathBar.max_value = health
	%HelathBar.value = health
	%InfoPanel/Info.text = "‚ù§Ô∏è%s\n‚öïÔ∏è%s üí™%s" % [health, STATS.heal, STATS.damage]
	%InfoPanel/SkillInfo.text = '–£–º–µ–Ω–∏–π: %s\n–†–∞–∑–≤–∏—Ç–∏–π: %s' % [skills_component.skills.size(), mastery_component.upgrades.size()]
	
func set_active(active: bool):
	active_component.set_active(active)

func take_damage(amount): 
	var old_health = health
	health -= amount
	health_changed.emit(old_health, health)
	if health <= 0:
		health_depleted.emit()
	update_unit_info()
		
func _on_actions_attacked():
	prints("DAMAGE", STATS.damage)
	take_damage(STATS.damage)

## –°–º–µ–Ω–∞ —Ü–µ–ª–∏ –ø–æ –∫–ª–∏–∫—É –ø–µ—Ä—Å–æ–Ω–∞–∂–∞
func _on_gui_input(event: InputEvent) -> void:
	if event.is_action_pressed('click'):
		$TargetComponent.toggle_target()
		
func serialize(): return $SerializerComponent.serialize()
func deserialize(data): return $SerializerComponent.deserialize(data)

## –°–∏–≥–Ω–∞–ª: –º–∞—Å—Ç–µ—Ä—Å—Ç–≤–æ –±—ã–ª–æ —É–ª—É—á—à–µ–Ω–æ —É –ø–µ—Ä—Å–æ–Ω–∞–∂–∞
func _on_mastery_component_mastery_upgraded() -> void:
	update_unit_info()

## –°–∏–≥–Ω–∞–ª - –∫–æ–≥–¥–∞ –ø–æ–≤—ã—à–∞–µ—Ç—Å—è —É—Ä–æ–≤–µ–Ω—å
func _on_level_component_level_upgraded() -> void:
	update_unit_info()
